torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -254,34 +276,75 @@ def _post_backward(self, *unused: Any):
             )
 
     def finalize_backward(self):
-        if self._sharded_state == ShardedState.UNSHARDED:
-            # Run post-backward here since the forward inputs did not require
-            # gradient, so the post-backward hook did not run
-            self._post_backward()
         if self._reduce_scatter_view_out_event is not None:
             torch.cuda.current_stream().wait_event(self._reduce_scatter_view_out_event)
             self._reduce_scatter_view_out_event = None
         self._training_state = TrainingState.IDLE
+        self._post_forward_indices.clear()
+        self.all_forward_grad_fns.clear()
+
+    def _prefetch_unshard(self):
+        if self._training_state == TrainingState.PRE_BACKWARD:
+            if not self._post_forward_indices:
+                # Can happen if running backward >1 time through this module
+                # (with `retain_graph=True`)
+                return
+            curr_index = self._post_forward_indices.pop()
+            if (target_index := curr_index - 1) < 0:
+                return
+            target_fsdp_param_group = self.comm_ctx.post_forward_order[target_index]
+            if any(
+                torch._C._will_engine_execute_node(grad_fn)  # type: ignore[attr-defined]
```

Comment:
`_will_engine_execute_node` is `O(1)`, so this check is linear in the number of forward outputs (that require gradient) that the target parameter group was involved in. This seems tolerable to me.